### PR TITLE
add several byzantine robust aggregators

### PR DESCRIPTION
**Pull request by private-mechanism**

### CODE REVIEW

**Comment by private-mechanism** [comment]
Add several Byzantine robust aggregators and modify corresponding auxiliary files.

File: federatedscope/core/workers/server.py [hunk 1]
```diff
@@ -151,6 +151,27 @@ def __init__(self,
                     for _ in range(self.model_num - 1)
                 ])
 
+        if self._cfg.aggregator.fltrust.use:
+            # set up a trainer for conducting global model training and
+            # model evaluation in server
+            assert self.model is not None
+            assert self.data is not None
+            self.trainer = get_trainer(model=self.model,
```
Comment (xieyxclack) [inline federatedscope/core/workers/server.py:159]: Why not merge this with line 133-151?

File: federatedscope/core/workers/server.py [hunk 2]
```diff
@@ -901,6 +926,27 @@ def eval(self):
             self.broadcast_model_para(msg_type='evaluate',
                                       filter_unseen_clients=False)
 
+    def _calculate_model_delta(self, init_model, updated_model):
+        model_delta = copy.deepcopy(self.model.state_dict())
+        for key in self.model.state_dict():
+            model_delta[key] = updated_model[key] - init_model[key]
+        return model_delta
+
+    def _global_trainer(self):
```
Comment (xieyxclack) [inline federatedscope/core/workers/server.py:935]: If it is only used for `Fltrust aggregator`, I think it should not be placed in `server.py`

**Review by xieyxclack — COMMENTED** [review]
God job, these aggregators are useful in developing FL algorithms, thank you. Some suggestions are given as inline comments. Note that the details of the implemented algorithms have not been carefully reviewed and would be reviewed ASAP!

File: federatedscope/core/configs/cfg_aggregator.py [hunk 3]
```diff
@@ -29,9 +40,12 @@ def extend_aggregator_cfg(cfg):
 
 def assert_aggregator_cfg(cfg):
 
-    if cfg.aggregator.byzantine_node_num == 0 and cfg.aggregator.krum.use:
-        logging.warning('Although krum aggregtion rule is applied, we found '
-                        'that cfg.aggregator.byzantine_node_num == 0')
+    if cfg.aggregator.byzantine_node_num == 0 and \
+            cfg.aggregator.robust_rule in \
+            ['krum', 'normbounding', 'median', 'trimmedmean', 'bulyan']:
```
Comment (xieyxclack) [inline federatedscope/core/configs/cfg_aggregator.py:45]: ‘fedavg’?

File: federatedscope/core/aggregators/median_aggregator.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import numpy as np
+from federatedscope.core.aggregators import ClientsAvgAggregator
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class MedianAggregator(ClientsAvgAggregator):
+    """
+    Implementation of median refers to `Byzantine-robust distributed
```
Comment (xieyxclack) [inline federatedscope/core/aggregators/median_aggregator.py:12]: please add refs here

File: federatedscope/core/auxiliaries/aggregator_builder.py [hunk 5]
```diff
@@ -87,8 +89,20 @@ def get_aggregator(method, model=None, device=None, online=False, config=None):
             return AsynClientsAvgAggregator(model=model,
                                             device=device,
                                             config=config)
-        elif config.aggregator.krum.use:
+        elif config.aggregator.robust_rule == 'krum':
```
Comment (xieyxclack) [inline federatedscope/core/auxiliaries/aggregator_builder.py:92]: FedAvg

**Pull request closed without merging**